File: ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/blocks/exec/ExecutableData.java
/*
 * Copyright 2020 ConsenSys AG.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */

package tech.pegasys.teku.datastructures.blocks.exec;

import static tech.pegasys.teku.util.config.Constants.BYTES_PER_LOGS_BLOOM;
import static tech.pegasys.teku.util.config.Constants.MAX_ETH1_TRANSACTIONS;

import com.google.common.base.MoreObjects;
import java.util.List;
import java.util.function.Function;
import org.apache.tuweni.bytes.Bytes;
import org.apache.tuweni.bytes.Bytes32;
import tech.pegasys.teku.datastructures.util.SpecDependent;
import tech.pegasys.teku.infrastructure.logging.LogFormatter;
import tech.pegasys.teku.infrastructure.unsigned.UInt64;
import tech.pegasys.teku.ssz.SSZTypes.Bytes20;
import tech.pegasys.teku.ssz.SSZTypes.SSZBackingList;
import tech.pegasys.teku.ssz.SSZTypes.SSZList;
import tech.pegasys.teku.ssz.backing.ListViewRead;
import tech.pegasys.teku.ssz.backing.VectorViewRead;
import tech.pegasys.teku.ssz.backing.containers.Container10;
import tech.pegasys.teku.ssz.backing.containers.ContainerType10;
import tech.pegasys.teku.ssz.backing.tree.TreeNode;
import tech.pegasys.teku.ssz.backing.type.BasicViewTypes;
import tech.pegasys.teku.ssz.backing.type.ListViewType;
import tech.pegasys.teku.ssz.backing.type.VectorViewType;
import tech.pegasys.teku.ssz.backing.view.BasicViews.ByteView;
import tech.pegasys.teku.ssz.backing.view.BasicViews.Bytes32View;
import tech.pegasys.teku.ssz.backing.view.BasicViews.UInt64View;
import tech.pegasys.teku.ssz.backing.view.ViewUtils;

public class ExecutableData
    extends Container10<
        ExecutableData,
        Bytes32View,
        Bytes32View,
        VectorViewRead<ByteView>,
        Bytes32View,
        UInt64View,
        UInt64View,
        Bytes32View,
        VectorViewRead<ByteView>,
        UInt64View,
        ListViewRead<Eth1Transaction>> {

  public static class ExecutableDataType
      extends ContainerType10<
          ExecutableData,
          Bytes32View,
          Bytes32View,
          VectorViewRead<ByteView>,
          Bytes32View,
          UInt64View,
          UInt64View,
          Bytes32View,
          VectorViewRead<ByteView>,
          UInt64View,
          ListViewRead<Eth1Transaction>> {

    public ExecutableDataType() {
      super(
          "ExecutableData",
          namedType("parent_hash", BasicViewTypes.BYTES32_TYPE),
          namedType("block_hash", BasicViewTypes.BYTES32_TYPE),
          namedType("coinbase", new VectorViewType<>(BasicViewTypes.BYTE_TYPE, Bytes20.SIZE)),
          namedType("state_root", BasicViewTypes.BYTES32_TYPE),
          namedType("gas_limit", BasicViewTypes.UINT64_TYPE),
          namedType("gas_used", BasicViewTypes.UINT64_TYPE),
          namedType("receipt_root", BasicViewTypes.BYTES32_TYPE),
          namedType(
              "logs_bloom", new VectorViewType<>(BasicViewTypes.BYTE_TYPE, BYTES_PER_LOGS_BLOOM)),
          namedType("difficulty", BasicViewTypes.UINT64_TYPE),
          namedType(
              "transactions",
              new ListViewType<>(Eth1Transaction.TYPE.get(), MAX_ETH1_TRANSACTIONS)));
    }

    public ListViewType<Eth1Transaction> getTransactionsType() {
      return (ListViewType<Eth1Transaction>) getFieldType9();
    }

    @Override
    public ExecutableData createFromBackingNode(TreeNode node) {
      return new ExecutableData(this, node);
    }
  }

  private ExecutableData(ExecutableDataType type, TreeNode backingNode) {
    super(type, backingNode);
  }

  public static final SpecDependent<ExecutableDataType> TYPE =
      SpecDependent.of(ExecutableDataType::new);

  public ExecutableData() {
    super(TYPE.get());
  }

  public ExecutableData(
      ExecutableDataType type,
      Bytes32 parent_hash,
      Bytes32 block_hash,
      Bytes20 coinbase,
      Bytes32 state_root,
      UInt64 gas_limit,
      UInt64 gas_used,
      Bytes32 receipt_root,
      Bytes logs_bloom,
      UInt64 difficulty,
      Iterable<Eth1Transaction> transactions) {
    super(
        type,
        new Bytes32View(parent_hash),
        new Bytes32View(block_hash),
        ViewUtils.createVectorFromBytes(coinbase.getWrappedBytes()),
        new Bytes32View(state_root),
        new UInt64View(gas_limit),
        new UInt64View(gas_used),
        new Bytes32View(receipt_root),
        ViewUtils.createVectorFromBytes(logs_bloom),
        new UInt64View(difficulty),
        ViewUtils.toListView(type.getTransactionsType(), transactions));
  }

  public ExecutableData(
      Bytes32 parent_hash,
      Bytes32 block_hash,
      Bytes20 coinbase,
      Bytes32 state_root,
      UInt64 gas_limit,
      UInt64 gas_used,
      Bytes32 receipt_root,
      Bytes logs_bloom,
      UInt64 difficulty,
      List<Eth1Transaction> transactions) {
    this(
        TYPE.get(),
        parent_hash,
        block_hash,
        coinbase,
        state_root,
        gas_limit,
        gas_used,
        receipt_root,
        logs_bloom,
        difficulty,
        transactions);
  }

  public Bytes32 getParent_hash() {
    return ((Bytes32View) get(0)).get();
  }

  public Bytes32 getBlock_hash() {
    return ((Bytes32View) get(1)).get();
  }

  public Bytes20 getCoinbase() {
    return new Bytes20(ViewUtils.getAllBytes(getAny(2)));
  }

  public Bytes32 getState_root() {
    return ((Bytes32View) get(3)).get();
  }

  public UInt64 getGas_limit() {
    return ((UInt64View) get(4)).get();
  }

  public UInt64 getGas_used() {
    return ((UInt64View) get(5)).get();
  }

  public Bytes32 getReceipt_root() {
    return ((Bytes32View) get(6)).get();
  }

  public Bytes getLogs_bloom() {
    return ViewUtils.getAllBytes(getAny(7));
  }

  public UInt64 getDifficulty() {
    return ((UInt64View) get(8)).get();
  }

  public SSZList<Eth1Transaction> getTransactions() {
    return new SSZBackingList<>(
        Eth1Transaction.class, getAny(9), Function.identity(), Function.identity());
  }

  @Override
  public String toString() {
    return MoreObjects.toStringHelper(this)
        .add("parent_hash", LogFormatter.formatHashRoot(getParent_hash()))
        .add("block_hash", LogFormatter.formatHashRoot(getBlock_hash()))
        .add("coinbase", getCoinbase().getWrappedBytes().slice(0, 8))
        .add("state_root", LogFormatter.formatHashRoot(getState_root()))
        .add("gas_limit", getGas_limit())
        .add("gas_used", getGas_used())
        .add("receipt_root", LogFormatter.formatHashRoot(getReceipt_root()))
        .add("logs_bloom", getLogs_bloom().slice(0, 8))
        .add("difficulty", getDifficulty())
        .toString();
  }
}

File: phase1/src/main/kotlin/tech/pegasys/teku/phase1/integration/datastructures/ExecutableData.kt
package tech.pegasys.teku.phase1.integration.datastructures

import org.apache.tuweni.units.bigints.UInt256
import tech.pegasys.teku.phase1.integration.getBasicValue
import tech.pegasys.teku.phase1.integration.ssz.SSZByteListImpl
import tech.pegasys.teku.phase1.integration.ssz.SSZByteVectorImpl
import tech.pegasys.teku.phase1.integration.ssz.SSZListImpl
import tech.pegasys.teku.phase1.integration.ssz.getListView
import tech.pegasys.teku.phase1.integration.wrapValues
import tech.pegasys.teku.phase1.onotole.ssz.Bytes
import tech.pegasys.teku.phase1.onotole.ssz.Bytes20
import tech.pegasys.teku.phase1.onotole.ssz.Bytes32
import tech.pegasys.teku.phase1.onotole.ssz.SSZByteList
import tech.pegasys.teku.phase1.onotole.ssz.SSZByteVector
import tech.pegasys.teku.phase1.onotole.ssz.SSZList
import tech.pegasys.teku.phase1.onotole.ssz.uint64
import tech.pegasys.teku.phase1.util.printRoot
import tech.pegasys.teku.ssz.backing.VectorViewRead
import tech.pegasys.teku.ssz.backing.tree.TreeNode
import tech.pegasys.teku.ssz.backing.type.BasicViewTypes
import tech.pegasys.teku.ssz.backing.type.ContainerViewType
import tech.pegasys.teku.ssz.backing.type.ListViewType
import tech.pegasys.teku.ssz.backing.type.VectorViewType
import tech.pegasys.teku.ssz.backing.view.AbstractImmutableContainer
import tech.pegasys.teku.ssz.backing.view.BasicViews
import tech.pegasys.teku.ssz.backing.view.ViewUtils

val MAX_TRANSACTIONS = 1uL shl 32
val MAX_TRANSACTION_SIZE = 1uL shl 32
val LOGS_BLOOM_SIZE = 256uL
val MAX_BYTES_PER_TRANSACTION_PAYLOAD = 1uL shl 32

class ExecutableData : AbstractImmutableContainer {

  val coinbase: Bytes20
    get() = getBasicValue(get(0))
  val state_root: Bytes32
    get() = getBasicValue(get(1))
  val gas_limit: uint64
    get() = getBasicValue(get(2))
  val gas_used: uint64
    get() = getBasicValue(get(3))
  val transactions: SSZList<Eth1Transaction>
    get() = SSZListImpl<Eth1Transaction, Eth1Transaction>(getAny(4)) { it }
  val receipt_root: Bytes32
    get() = getBasicValue(get(5))
  val logs_bloom: SSZByteVector
    get() = SSZByteVectorImpl(getAny<VectorViewRead<BasicViews.ByteView>>(6))
  val block_hash: Bytes32
    get() = getBasicValue(get(7))
  val difficulty: uint64
    get() = getBasicValue(get(8))

  constructor(
    type: ContainerViewType<out AbstractImmutableContainer>,
    backingNode: TreeNode
  ) : super(type, backingNode)

  constructor(
    coinbase: Bytes20,
    state_root: Bytes32,
    gas_limit: uint64,
    gas_used: uint64,
    transactions: List<Eth1Transaction>,
    receipt_root: Bytes32,
    logs_bloom: Bytes,
    block_hash: Bytes32,
    difficulty: uint64
  ) : super(
    TYPE,
    *wrapValues(
      coinbase,
      state_root,
      gas_limit,
      gas_used,
      getListView(Eth1Transaction.TYPE, MAX_TRANSACTIONS, transactions) { it },
      receipt_root,
      logs_bloom,
      block_hash,
      difficulty
    )
  )

  constructor() : super(TYPE)

  companion object {

    val TYPE = ContainerViewType(
      listOf(
        BasicViewTypes.BYTES32_TYPE,
        BasicViewTypes.BYTES32_TYPE,
        BasicViewTypes.UINT64_TYPE,
        BasicViewTypes.UINT64_TYPE,
        ListViewType<Eth1Transaction>(Eth1Transaction.TYPE, MAX_TRANSACTIONS.toLong()),
        BasicViewTypes.BYTES32_TYPE,
        VectorViewType<BasicViews.ByteView>(BasicViewTypes.BYTE_TYPE, LOGS_BLOOM_SIZE.toLong()),
        BasicViewTypes.BYTES32_TYPE,
        BasicViewTypes.UINT64_TYPE
      ),
      ::ExecutableData
    )
  }

  override fun toString(): String {
    return "(" +
        "coinbase=${coinbase}, " +
        "state_root=${printRoot(state_root)}, " +
        "gas_limit=${gas_limit}, " +
        "gas_used=${gas_used}, " +
        "transactions=${transactions.size}, " +
        "receipt_root=${printRoot(receipt_root)}, " +
        "logs_bloom=${
          ViewUtils.getAllBytes((logs_bloom as SSZByteVectorImpl).view).slice(0, 10)
        }, " +
        "block_hash=${printRoot(block_hash)}, " +
        "difficulty=$difficulty)"
  }
}


class Eth1Transaction : AbstractImmutableContainer {

  val nonce: uint64
    get() = getBasicValue(get(0))
  val gas_price: UInt256
    get() = getBasicValue(get(1))
  val gas_limit: uint64
    get() = getBasicValue(get(2))
  val recipient: Bytes20
    get() = getBasicValue(get(3))
  val value: UInt256
    get() = getBasicValue(get(4))
  val input: SSZByteList
    get() = SSZByteListImpl(getAny(5))
  val v: UInt256
    get() = getBasicValue(get(6))
  val r: UInt256
    get() = getBasicValue(get(7))
  val s: UInt256
    get() = getBasicValue(get(8))

  constructor(
    type: ContainerViewType<out AbstractImmutableContainer>,
    backingNode: TreeNode
  ) : super(type, backingNode)

  constructor(
    nonce: uint64,
    gas_price: UInt256,
    gas_limit: uint64,
    recipient: Bytes20,
    value: UInt256,
    input: Bytes,
    v: UInt256,
    r: UInt256,
    s: UInt256
  ) : super(
    TYPE,
    *wrapValues(
      nonce,
      gas_price,
      gas_limit,
      recipient,
      value,
      getListView(
        BasicViewTypes.BYTE_TYPE,
        MAX_BYTES_PER_TRANSACTION_PAYLOAD,
        input.toArray().toList()
      ) { BasicViews.ByteView(it) },
      v,
      r,
      s
    )
  )

  constructor() : super(TYPE)

  companion object {

    val TYPE = ContainerViewType(
      listOf(
        BasicViewTypes.UINT64_TYPE,
        BasicViewTypes.BYTES32_TYPE,
        BasicViewTypes.UINT64_TYPE,
        BasicViewTypes.BYTES32_TYPE,
        BasicViewTypes.BYTES32_TYPE,
        ListViewType<BasicViews.ByteView>(
          BasicViewTypes.BYTE_TYPE,
          MAX_BYTES_PER_TRANSACTION_PAYLOAD.toLong()
        ),
        BasicViewTypes.BYTES32_TYPE,
        BasicViewTypes.BYTES32_TYPE,
        BasicViewTypes.BYTES32_TYPE
      ),
      ::Eth1Transaction
    )
  }
}